### PR TITLE
New shape helper reorg for Conv, MaxPool, and AvgPool

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Next batch of changes for Conv, MaxPool, and AvgPool

Added facilities to make code more concise. Top shape helper now has this
```C++
  void computeShapeAndAssertOnFailure();
```
to reduce the amount of `if (failure()...)` boilerplate code.

Also, when no builder is passed, it defaults to analysis, further reducing the boiler plate code.

### CODE REVIEW

**Review by tungld — APPROVED** [review]
LGTM!

File: src/Transform/ONNX/ONNXDimAnalysis.cpp [hunk 1]
```diff
@@ -107,16 +107,37 @@ void exploreSameInputDims(const onnx_mlir::DimAnalysis::DimT &dim, ONNX_OP op,
   findAndAddSameDim(qmOuputIE, op.getOperation()->getOperands(), sameDims);
 }
 
+// _xxx postfix for new shape helper interface
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Transform/ONNX/ONNXDimAnalysis.cpp:110]: yes, eventually I will remove the other one and rename away the `_xxx`. It was an easy pattern to match.

File: src/Dialect/ONNX/ONNXOps/NewShapeHelper.hpp [hunk 2]
```diff
@@ -60,54 +60,62 @@ struct NewONNXOpShapeHelper {
...
+      IndexExprScope *scope);      /* Install local scope if null. */
   virtual ~NewONNXOpShapeHelper();
 
-  // Every child class is expected to create a computeShape with the following
+  // Every leaf class is expected to create a computeShape with the following
   // signature. This method is responsible to compute at a minimum the output
   // dims.
   virtual mlir::LogicalResult computeShape() = 0;
+  // Compute shape and assert on failure.
+  void computeShapeOrAssert();
```
[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Dialect/ONNX/ONNXOps/NewShapeHelper.hpp:103]: got it, that makes sense.

[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Dialect/ONNX/ONNXOps/NewShapeHelper.hpp]: changed to `computeShapeAndAssertOnFailure`

**Pull request merged**